### PR TITLE
Add support for batching

### PR DESCRIPTION
**Pull request by bbalser**

### CODE REVIEW

File: lib/etl.ex [hunk 1]
```diff
@@ -52,6 +52,11 @@ defmodule Etl do
     Etl.Pipeline.set_broadcast(pipeline, opts)
   end
 
+  @spec batch(Etl.Pipeline.t(), keyword) :: Etl.Pipeline.t()
```
Comment (jeffgrunewald) [inline lib/etl.ex:49]: Any reason not to just defdelegate these?

File: lib/etl/message.ex [hunk 2]
```diff
@@ -18,6 +18,10 @@ defmodule Etl.Message do
     %Etl.Message{message | data: update_function.(data)}
   end
 
+  def add_metadata(%__MODULE__{metadata: metadata} = message, key, term) do
```
Comment (jeffgrunewald) [inline lib/etl/message.ex]: Does this function need a name that indicates it only accepts new metadata, not updating existing keys?

File: lib/etl/stage/batcher.ex [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      |> Map.put(:queue, :queue.new())
+
+    {:producer_consumer, state}
+  end
+
+  def handle_events(events, _from, state) do
+    cancel_timer(state.timer)
+
+    queue =
+      for event <- events, reduce: state.queue do
```
Comment (jeffgrunewald) [inline lib/etl/stage/batcher.ex:30]: You saw Sean’s slack post too ;)

File: lib/etl/stage/batcher.ex [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        q -> :queue.in(event, q)
+      end
+
+    {outgoing, queue} =
+      case state.batch_size <= :queue.len(queue) do
```
Comment (jeffgrunewald) [inline lib/etl/stage/batcher.ex]: For some reason I really struggled with this logic. I think it reads easier if you switch it to say `case queue_len > batch_size`

File: test/support/batched_consumer.ex [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        Map.from_struct(state)
+        |> Map.merge(Map.new(opts))
+
+      {:automatic, state}
+    end
+
+    def handle_subscribe(_, _, _, state) do
+      {:automatic, state}
+    end
+
```
Comment (jeffgrunewald) [inline test/support/batched_consumer.ex]: Github flagged this as unused code although it didn’t actually fail any tests.

File: lib/etl/stage/batcher.ex (hunk 3 above)
Comment (bbalser) [reply lib/etl/stage/batcher.ex:30]: I did and thought I would give it a try.

File: lib/etl/message.ex (hunk 2 above)
Comment (bbalser) [reply lib/etl/message.ex]: The name does indicate that to me, but obviously not everybody, what about "put_new_metadata"?

Comment (jeffgrunewald) [reply lib/etl/message.ex]: Works for me. Or `add_new_metadata` or `put_new_meta`; i like your explicit mirror of the other map/keyword functions though.

File: lib/etl/stage/batcher.ex (hunk 3 above)
Comment (jeffgrunewald) [reply lib/etl/stage/batcher.ex:30]: i like it; when a `for` makes sense i think it's more expressive of the data flow than a reduce.

File: lib/etl/stage/batcher.ex (hunk 4 above)
Comment (bbalser) [reply lib/etl/stage/batcher.ex]: I read it over and it made more sense to me the way it is, so decided to pull the boolean expression out to another function.

**Review by jeffgrunewald — APPROVED** [review]

**Pull request merged**